src/edu/stanford/dlss/was/WasapiDownloader.java

```diff
@@ -44,10 +46,81 @@ void downloadSelectedWarcs() throws IOException {
       WasapiCrawlSelector crawlSelector = new WasapiCrawlSelector(wasapiRespList);
       for (Integer crawlId : desiredCrawlIds(crawlSelector)) {
         for (WasapiFile file : crawlSelector.getFilesForCrawl(crawlId)) {
-          // TODO:  make a separate method for downloading individual file?
-          System.out.println("We will eventually download " + file.toString());
+          downloadAndValidateFile(file);
+        }
+      }
+    }
+  }
+
+  // package level method for testing
+  WasapiConnection getWasapiConn() throws IOException {
+    if (wasapiConn == null)
+      wasapiConn = new WasapiConnection(new WasapiClient(settings));
+    return wasapiConn;
+  }
+
+  // package level method for testing
+  @SuppressWarnings("checkstyle:MethodLength")
+  void downloadAndValidateFile(WasapiFile file) throws NoSuchAlgorithmException {
+    String fullFilePath = prepareOutputLocation(file);
+    if (fullFilePath == null) {
+      System.err.println("fullFilePath is null - can't retrieve file");
+      return;
+    }
+    int attempts = 0;
+    boolean notValidated = true;
+    do {
+      attempts++;
+      try {
+        // System.out.println("DEBUG: trying to get " + file.getLocations()[0]);
+        boolean downloadSuccess = getWasapiConn().downloadQuery(file.getLocations()[0], fullFilePath);
+        if (downloadSuccess && checksumValidate("md5", file, fullFilePath)) {
+          System.out.println("file retrieved successfully: " + file.getLocations()[0]);
+          notValidated = false; // break out of loop
         }
+      } catch (HttpResponseException e) {
+        System.err.println("ERROR: HttpResponseException downloading file (will not retry): " + file.getLocations()[0]);
+        System.err.println(" HTTP ResponseCode was " + e.getStatusCode());
+        // e.printStackTrace(System.err);
```

Comment:
when I was writing tests, the stack trace seemed superfluous, and provided no real info for debugging that we didn't already have.  I will remove this and the next.